File: src/engine/configuration/options.ts
import { OptionName } from 'sonolus-core'
import { defineOptions } from 'sonolus.js'

export const options = defineOptions({
    isAutoplay: {
        name: OptionName.AutoPlay,
        scope: 'Bandori',
        standard: true,
        type: 'toggle',
        def: 0,
    },
    isNoteSpeedRandom: {
        name: OptionName.RandomNoteSpeed,
        scope: 'Bandori',
        standard: true,
        type: 'toggle',
        def: 0,
    },
    isStrictJudgment: {
        name: OptionName.StrictJudgment,
        scope: 'Bandori',
        standard: true,
        type: 'toggle',
        def: 0,
    },
    speed: {
        name: OptionName.LevelSpeed,
        standard: true,
        type: 'slider',
        def: 1,
        min: 0.5,
        max: 2,
        step: 0.05,
        display: 'percentage',
    },
    noteSpeed: {
        name: OptionName.NoteSpeed,
        scope: 'Bandori',
        type: 'slider',
        def: 5,
        min: 1,
        max: 15.9,
        step: 0.1,
        display: 'number',
    },
    noteSize: {
        name: OptionName.NoteSize,
        scope: 'Bandori',
        type: 'slider',
        def: 1,
        min: 0.1,
        max: 2,
        step: 0.05,
        display: 'percentage',
    },
    noteEffectSize: {
        name: OptionName.NoteEffectSize,
        scope: 'Bandori',
        type: 'slider',
        def: 1,
        min: 0.1,
        max: 2,
        step: 0.05,
        display: 'percentage',
    },
    connectorAlpha: {
        name: OptionName.ConnectorTransparency,
        scope: 'Bandori',
        type: 'slider',
        def: 0.8,
        min: 0.1,
        max: 1,
        step: 0.05,
        display: 'percentage',
    },
    stageCover: {
        name: OptionName.VerticalStageCover,
        scope: 'Bandori',
        type: 'slider',
        def: 0,
        min: 0,
        max: 1,
        step: 0.05,
        display: 'percentage',
    },
    isMirrorEnabled: {
        name: OptionName.MirrorLevel,
        type: 'toggle',
        def: 0,
    },
    isSimLineEnabled: {
        name: OptionName.SimultaneousLineVisibility,
        scope: 'Bandori',
        type: 'toggle',
        def: 1,
    },
    isSFXEnabled: {
        name: OptionName.SFX,
        scope: 'Bandori',
        type: 'toggle',
        def: 1,
    },
    isNoteEffectEnabled: {
        name: OptionName.NoteEffect,
        scope: 'Bandori',
        type: 'toggle',
        def: 1,
    },
    isLaneEffectEnabled: {
        name: OptionName.LaneEffect,
        scope: 'Bandori',
        type: 'toggle',
        def: 1,
    },
    isSlotEffectEnabled: {
        name: OptionName.SlotEffect,
        scope: 'Bandori',
        type: 'toggle',
        def: 1,
    },
    isStageAspectRatioLocked: {
        name: OptionName.LockStageAspectRatio,
        scope: 'Bandori',
        type: 'toggle',
        def: 1,
    },
})

File: engine/configuration/options.cpp
using namespace std;

class Options {
    public:

    int autoplay = 0;
    int strictJudge = 1;
    int mirror = 2;
    int noteSpeed = 3;
    int noteSize = 4;
    int syncLine = 5;
    int lockAspectRatio = 6;
    int stage = 7;
}Options;

auto options = defineOptions<class Options>({
    {
        .name = NameText.AutoPlay,
        .scope = Scope,
        .standard = 1,
        .type = OptionType.Toggle,
        .def = 0,
    }, {
        .name = NameText.StrictJudgment,
        .scope = Scope,
        .standard = 1,
        .type = OptionType.Toggle,
        .def = 0,
    }, {
        .name = NameText.MirrorLevel,
        .scope = Scope,
        .standard = 0,
        .type = OptionType.Toggle,
        .def = 0
    }, {
        .name = NameText.NoteSpeed,
        .scope = Scope,
        .standard = 0,
        .type = OptionType.Slider,
        .def = 5,
        .min = 1,
        .max = 50,
        .step = 0.1,
    }, {
        .name = NameText.NoteSize,
        .scope = Scope,
        .standard = 0,
        .type = OptionType.Slider,
        .def = 1,
        .min = 0.5,
        .max = 2.0,
        .step = 0.1,
        .unit = UnitText.Percentage,
    }, {
        .name = NameText.SimultaneousLineVisibility,
        .scope = Scope,
        .standard = 1,
        .type = OptionType.Toggle,
        .def = 1,
    }, {
        .name = NameText.LockStageAspectRatio,
        .scope = Scope,
        .standard = 1,
        .type = OptionType.Toggle,
        .def = 0,
    }, {
        .name = NameText.StageVisibility,
        .scope = Scope,
        .standard = 1,
        .type = OptionType.Toggle,
        .def = 1,
    }
});